text/000-sort-variable-unification.md

```diff
@@ -0,0 +1,77 @@
+
+The first part of this CEP is implemented in [Coq PR 16903](https://github.com/coq/coq/pull/16903).
+
+# Motivation
+
+There are several underlying motivations for this CEP.
+
+- Immediate fix: alleviate the longstanding issue of having to explicitly
+annotate variables with `Prop` to make the unification go through.
+- Short term: fix the `SProp` mess in the implementation of kernel and unification.
+- Long term: enhance the modularity of Coq by allowing terms to be polymorphic
+over sorts like `Prop` / `Type` / `SProp`, a.k.a. stop triplicating everything.
+
+The first point is adressed directly in this CEP, the two other ones are made
+possible after the CEP is implemented. We detail the issues below.
+
+As of Coq 8.17, every time one creates a fresh sort, it is implicitly living in the `Type` hierarchy. This has unfortunate consequences, like the fact that `fun A => A /\ A -> A` typechecks but `fun A => A -> A /\ A` does not. Since the arrow is typechecked left-to-right, the first one forces `A : Prop` but the second one forces `A` to be a fresh sort, hence a `Type`. In that latter case you have to explicitly write the Prop annotation for the pretyper to be happy. Similar issues arise from other asymmetrical constructs like equality.
+
+The introduction of `SProp` makes the problem even more salient. Long story short, the historical universe unification algorithm in Coq basically assumes that everything is a `Type`, and the reason `Prop` ⊆ `Type` is not so much about kernel expressivity but rather that otherwise it would be too painful to write code using `Prop`. `SProp` has the same issue, so for easiness of writing `SProp`-based code and not confusing unification, @SkySkimmer introduced a horrible hack where we pretend `SProp` ⊆ `Type`, i.e. the dreaded cumulative `SProp` flag. The problem is that `SProp` also taints the terms in the relevance marks, so in the end one gets nonsense terms in the kernel that needs to fix them on the fly. This is the usual curse of Coq technical debt: you fix issues in the upper layers by introducing workarounds in the kernel. This makes it essentially impossible to use `SProp` at scale.
+
+# Detailed design
+
+This CEP introduces a trivial idea to sort out the mess. Namely, in addition to the well-known universe *levels* one now also has access to sorts *qualities*. While levels are basically integers, qualities are the nature of the sort itself, i.e. `Prop`, `Type` or `SProp`. Since we want to have unification on this first-order signature, we also freely add variables. They are just tracked in the evarmap in a way similar to levels. The most generic type is thus `QSort (α, u)` where `α` is a sort quality and `u` an algebraic universe.
+
+Since this affects pervasive data structures, we need to change the kernel definitions. But similarly to evars, the kernel currently does not know anything about undefined sort variables, apart for their relevance which is needed for conversion.
```

Comment:
Nit:
```suggestion
Since this affects pervasive data structures, we need to change the kernel definitions. But similarly to evars, the kernel currently does not know anything about undefined sort variables, apart from their relevance which is needed for conversion.
```